src/uci_options.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include "parsing/input_parsing.h"
+
+#include "fmt/base.h"
+#include <cstdint>
+#include <functional>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace ucioption {
+
+/* uci types for easier creation */
+using string = std::string;
+using check = bool;
+using spin = uint64_t;
+
+template<typename T>
+struct Storage {
+    T value;
+    const T defaultValue;
+    const std::function<void(const T&)> callback;
+};
+
+using Variant = std::variant<
```

Comment:
I really like this as a solution to class inheritance plus dynamic sizing, without heap allocation :)

Reply:
It's pretty cool, yes! Can lead to some pretty complicated implementations if you don't take care though.

Reply:
Just a heads-up in regards to `dynamic sizing`:

This size allocated is the biggest of the variants. So it's not actually memory efficient if you were to have bigger types. For our sake it really doesn't matter though. But something to be aware of :)